packages/tokenbank_v1/contracts/impl/BaseWallet.sol

```diff
@@ -0,0 +1,242 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../iface/Wallet.sol";
+import "../iface/Module.sol";
+
+import "../lib/ERC20.sol";
+import "../lib/NamedAddressSet.sol";
+import "../lib/ReentrancyGuard.sol";
+
+
+/// @title BaseWallet
+/// @dev This contract provides basic implementation for a Wallet.
+///
+/// @author Daniel Wang - <daniel@loopring.org>
+///
+/// The design of this contract is inspired by Argent's contract codebase:
+/// https://github.com/argentlabs/argent-contracts
+contract BaseWallet is Wallet, NamedAddressSet, ReentrancyGuard
+{
+    string private constant MODULE = "__MODULE__";
+    string private constant ERC20_TRANSFER = "transfer(address,uint256)";
+
+    mapping (bytes4  => address) internal methodToModule;
+
+    /// @dev Emitted when the wallet received Ether.
+    event Received(
+        address indexed sender,
+        uint    value,
+        bytes   data
+    );
+
+    modifier onlyOwner
+    {
+        require(msg.sender == owner, "NOT_A_OWNER");
+        _;
+    }
+
+    modifier onlyModule
+    {
+        require(isAddressInSet(MODULE, msg.sender), "NOT_A_MODULE");
+        _;
+    }
+
+    function init(
+        address            _owner,
+        address[] calldata _modules
+        )
+        external
+        nonReentrant
+    {
+        require(owner == address(0) && numAddressesInSet(MODULE) == 0, "INITIALIZED_ALREADY");
+        require(_owner != address(0), "ZERO_ADDRESS");
+        require(_modules.length > 0, "EMPTY_MODULES");
+
+        owner = _owner;
+
+        emit Initialized(owner);
+
+        for(uint i = 0; i < _modules.length; i++) {
+            address module = _modules[i];
+            require(module != address(0), "NULL_MODULE");
```

Comment:
Seems like this does the same as `addModule` so it may not be a bad idea to create an `addModuleInternal` so the logic can be shared.

Reply:
done.